### PR TITLE
Create a Tempo Bundle

### PR DESCRIPTION
**Pull request by michaeldmitry**
Create a Tempo bundle with tls-overlay.

## Testing instructions
Run
```
tox -e integration
```

### CODE REVIEW

File: .github/workflows/ci.yaml
Comment (mmkay) [inline .github/workflows/ci.yaml]: I wonder if these could be common for our HA charms and part of shared workflows: https://github.com/canonical/observability/tree/main/.github/workflows

File: CONTRIBUTING.md
Comment (mmkay) [inline CONTRIBUTING.md]: `s3proxy` -> `s3-integrator`

File: README.md
Comment (mmkay) [inline README.md:35]: Might also be worth noting in the `README` that the bundle won't work without configuring s3-integrator to point at something s3-compatible, be it s3 credentials or another charm that does the s3 work like minio.

File: bundle.yaml.j2
Comment (mmkay) [inline bundle.yaml.j2:18]: You'll also need `resources: nginx-image: (...) nginx-prometheus-exporter-image:` as you can see here: https://github.com/canonical/tempo-coordinator-k8s-operator/pull/4/files#diff-8af267359854d30b091fde763088e458bf27c8fedc056895f3104ef05b61f4b5R25

File: overlays/tls-overlay.yaml
Comment (mmkay) [inline overlays/tls-overlay.yaml:8]: I wonder if we should have it for the official bundle or leave it to the end users to configure.

File: render_bundle.py
Comment (mmkay) [inline render_bundle.py:1]: looks a bit like something common to all bundles that could land in observability repo?

**Review by mmkay — COMMENTED** [review]
Nice one! First round of comments but if @PietroPasotti can also have a look that'd be great.

File: CONTRIBUTING.md
Comment (PietroPasotti) [inline CONTRIBUTING.md]: mmm this is not correct, is it? By default the bundle points to charmhub charms so it will deploy tempo from stable/edge...? IIRC you need to go and change the charm name in the rendered bundle to point it to a local path for this to work.

File: CONTRIBUTING.md
Comment (PietroPasotti) [inline CONTRIBUTING.md]:
```suggestion
This bundle is comprised of two charms:
```

File: CONTRIBUTING.md
Comment (PietroPasotti) [inline CONTRIBUTING.md]:
```suggestion
- [tempo-coordinator-k8s](https://github.com/canonical/tempo-coordinator-k8s-operator/)
- [tempo-worker-k8s](https://github.com/canonical/tempo-worker-k8s-operator/)
```

File: CONTRIBUTING.md
Comment (PietroPasotti) [inline CONTRIBUTING.md]:
```suggestion
- [s3-integrator](https://github.com/canonical/s3-integrator)
```

File: CONTRIBUTING.md
Comment (PietroPasotti) [inline CONTRIBUTING.md]: tox -e integration --notest will not create any model.

 ```suggestion
Now `juju add-model foo` and deploy the bundle with `juju deploy ./bundle.yaml`
```

do we need `--trust` too perhaps?

Comment (PietroPasotti) [reply CONTRIBUTING.md]: also add a link to a minio deployment guide as @mmkay suggested

File: CONTRIBUTING.md
Comment (PietroPasotti) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md
Comment (PietroPasotti) [inline CONTRIBUTING.md]:
```suggestion
To keep the model and applications running after the test suite has exited:
```

File: README.md
Comment (PietroPasotti) [inline README.md:56]: I wonder if we should keep calling this the 'monolithic mode', to avoid suggesting that we also have a non-scaling one.
Much like we internally remap `all` to `scaling-single-binary`.
Perhaps add a note that the official docs call it differently, but to us this is just "the monolithic mode"? Or will that be just confusing?
@simskij WDYT?

File: README.md
Comment (PietroPasotti) [inline README.md:67]: at line 62, please add a note that what makes this mode 'monolithic' is that the worker application is configured with `role=all`.

File: README.md
Comment (PietroPasotti) [inline README.md]: perhaps I would write the app names in this listing too, separate from their scale, so the user will know what to expect to see in `juju status`.
How about:

```
The bundle will deploy:

- a `tempo` application with 1 unit (`tempo-coordinator-k8s` charm)
- a `tempo-worker` application with 3 units (`tempo-worker-k8s` charm), configured with `role=all`
- an `s3-integrator` application with 1 unit (`s3-integrator` charm)
```

File: README.md
Comment (PietroPasotti) [inline README.md]:
```suggestion
The bundle can also be rendered to deploy Tempo in its [microservices mode](https://grafana.com/docs/tempo/latest/setup/deployment/#microservices-mode), which runs each one of the required roles in distinct processes. Each required role is assigned to one worker unit to achieve the minimal deployment for the cluster to be considered consistent. Minimal microservices mode is the way to deploy Grafana Tempo in a distributed architecture without high availability.
```

File: README.md
Comment (PietroPasotti) [inline README.md]:
```suggestion
Pass the `--mode=minimal-microservices` argument to render the bundle for this mode:
```

File: README.md
Comment (PietroPasotti) [inline README.md:88]: please rephrase this listing following the same format as suggested for the previous one (include app names). I'd call the worker nodes `tempo-<role>`, omitting the 'worker'

File: README.md
Comment (PietroPasotti) [inline README.md]:
```suggestion
Finally, the bundle can be rendered to deploy the 'recommended' amount of nodes for its microservices mode. In this scenario, each worker node is scaled up to the recommended number of units. This is the recommended way to deploy a highly available Grafana Tempo for production environments.
```

Should we add a note/link explaining who is doing the recommendation? Is this according to some official grafana guide, helm chart, or is this our expert opinion?

File: README.md
Comment (PietroPasotti) [inline README.md]: rephrase as above

File: README.md
Comment (PietroPasotti) [inline README.md:109]: juju add-model tempo-prod

File: README.md
Comment (PietroPasotti) [inline README.md]: also reformat as above

File: bundle.yaml.j2
Comment (PietroPasotti) [inline bundle.yaml.j2]:
```suggestion
  tempo-{{ role }}:
```

File: bundle.yaml.j2
Comment (PietroPasotti) [inline bundle.yaml.j2]:
```suggestion
  - tempo-{{ role }}:tempo-cluster
```

File: render_bundle.py
Comment (PietroPasotti) [reply render_bundle.py:1]: observability-bundles perhaps, to also have the GH CI automation stuff?

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
Looking good! Bunch of recommendations here and there, nothing major.

File: render_bundle.py
Comment (michaeldmitry) [reply render_bundle.py:1]: @PietroPasotti  @mmkay 
Is observability-bundles an existing repo? (Can't find it maybe due to permissions)
If you mean in a new repo, I feel we could just put them in one of our existing.
CI Stuff: its mostly the same we have in  https://github.com/canonical/observability/tree/main/.github/workflows, so we can just make use of them.
render_bundle.py: maybe this could lie in https://github.com/canonical/observability/ or https://github.com/canonical/cos-lib/?
WDYT

Comment (mmkay) [reply render_bundle.py:1]: @michaeldmitry: there's cos-lite-bundle here: https://github.com/canonical/cos-lite-bundle but I don't think we have a bundle parent repo yet. I still think we could use `observability` as this is where we have the workflows.

File: overlays/tls-overlay.yaml
Comment (michaeldmitry) [reply overlays/tls-overlay.yaml:8]: I see your point.
I thought of keeping it coherent with what we have in https://github.com/canonical/cos-lite-bundle/blob/main/overlays/tls-overlay.yaml#L7

File: README.md
Comment (michaeldmitry) [reply README.md]: Its still a TODO item I guess to find out what does the recommended deployment look like

File: README.md
Comment (michaeldmitry) [reply README.md:56]: IMO, I think it would be confusing. I think we clearly say that we only have 3 modes of operation and we just mention what is that mean in the official docs.

File: README.md
Comment (michaeldmitry) [reply README.md]: I see grafana have their helm chart for tempo with these values https://github.com/grafana/helm-charts/blob/main/charts/tempo-distributed/values.yaml
Guessing we can have this as our initial recommendation at least until we have a counter argument.

File: README.md
Comment (simonaronsson) [reply README.md:56]: I think I'm with Michael on this one.

File: overlays/tls-overlay.yaml
Comment (mmkay) [reply overlays/tls-overlay.yaml:8]: Fair enough - it seems okay to stay coherent.

File: render_bundle.py
Comment (michaeldmitry) [reply render_bundle.py:1]: https://github.com/canonical/observability/pull/189

File: .github/workflows/pull-request.yaml
Comment (lucabello) [inline .github/workflows/pull-request.yaml]: Just noting that this should probably be changed to `@main` before merging :)

**Review by lucabello — APPROVED** [review]

**Review by PietroPasotti — APPROVED** [review]

**Pull request merged**
